src/workspaces/common/state/useWorkspaceStatePolling.ts

```diff
@@ -0,0 +1,118 @@
+
+const updateWorkspace = <T extends Workspace>(
+  workspace: T,
+  workspaceId: string,
+  state: WorkspaceState,
+  errorMessage?: string
+): T => {
+  if (workspace.workspace?.workspaceId === workspaceId) {
+    const update = _.cloneDeep(workspace);
+    update.workspace.state = state;
+    update.workspace.errorMessage = errorMessage;
+    return update;
+  }
+  return workspace;
+};
+
+const checkWorkspaceState = async (workspace: Workspace, abort: () => void, signal: AbortSignal) => {
+  const doUpdate = (abort: () => void, workspace: Workspace, state: WorkspaceState, errorMessage?: string) => {
+    const workspaceId = workspace.workspace.workspaceId;
+    workspacesStore.update((wsList) => updateWorkspacesList(wsList, workspaceId, state, errorMessage));
+    workspaceStore.update((ws) => {
+      if (ws) {
+        return updateWorkspace(ws, workspaceId, state, errorMessage);
+      }
+      return undefined;
+    });
+    abort();
+  };
+
+  const startingState = workspace.workspace.state;
+  try {
+    const wsResp: Workspace = await Ajax(signal)
+      .Workspaces.workspace(workspace.workspace.namespace, workspace.workspace.name)
+      .details(['workspace.state', 'workspace.errorMessage']);
+    const state = wsResp.workspace.state;
+
+    if (!!state && state !== startingState) {
+      doUpdate(abort, workspace, state, wsResp.workspace.errorMessage);
+    }
+  } catch (error) {
+    // for workspaces that are being deleted, the details endpoint will return a 404 when the deletion is complete
+    if (startingState === 'Deleting' && error instanceof Response && error.status === 404) {
+      doUpdate(abort, workspace, 'Deleted');
+    } else {
+      console.error(`Error checking workspace state for ${workspace.workspace.name}: ${JSON.stringify(error)}`);
+    }
+  }
+};
+
+const updatingStates: WorkspaceState[] = ['Deleting', 'Cloning', 'CloningContainer'];
+
+export const useWorkspaceStatePolling = (workspaces: Workspace[], status: LoadedState<Workspace[]>['status']) => {
+  // we have to do the signal/abort manually instead of with useCancelable so that the it can be cleaned up in the
+  // this component's useEffect, instead of the useEffect in useCancelable
+  const [controller, setController] = useState(new window.AbortController());
+  const abort = () => {
+    controller.abort();
+    setController(new window.AbortController());
+  };
+
+  useEffect(() => {
+    const iterateUpdatingWorkspaces = async () => {
+      if (status === 'Ready') {
+        const updatingWorkspaces = _.filter((ws) => _.contains(ws?.workspace?.state, updatingStates), workspaces);
+        for (const ws of updatingWorkspaces) {
+          await checkWorkspaceState(ws, abort, controller.signal);
+        }
+      }
+    };
+
+    pollWithCancellation(() => iterateUpdatingWorkspaces(), 30000, false, controller.signal);
+    return () => {
+      abort();
+    };
+    //  eslint-disable-next-line react-hooks/exhaustive-deps
```

Comment:
I find this comment often corresponds with subtle bugs. Is it possible to get rid of this?

Reply:
It appears that switching the abort controller from `useState` to `useRef` fixes the warning this was silencing - so yes!